x/ccv/provider/keeper/permissionless.go

```diff
@@ -0,0 +1,836 @@
+	store := ctx.KVStore(k.storeKey)
+
+	consumerIds, err := k.GetConsumersToBeLaunched(ctx, spawnTime)
+	if err != nil {
+		return err
+	}
+	consumerIds = append(consumerIds, consumerId)
+
+	// sort so that we avoid getting a consumer id in front of everyone else and delaying the appending of a chain
+	sort.Slice(consumerIds, func(i, j int) bool {
+		return consumerIds[i] < consumerIds[j]
+	})
+
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	err = enc.Encode(consumerIds)
+	if err != nil {
+		return err
+	}
+
+	store.Set(types.SpawnTimeToConsumerIdsKey(spawnTime), buf.Bytes())
+	return nil
+}
+
+// RemoveConsumerFromToBeLaunchedConsumers
+func (k Keeper) RemoveConsumerFromToBeLaunchedConsumers(ctx sdk.Context, consumerId string, spawnTime time.Time) error {
+	store := ctx.KVStore(k.storeKey)
+
+	consumerIds, err := k.GetConsumersToBeLaunched(ctx, spawnTime)
+	if err != nil {
+		return err
+	}
+
+	if len(consumerIds) == 0 {
+		return fmt.Errorf("no consumer ids for spawn time: %s", spawnTime.String())
+	}
+
+	// find the index of the consumer we want to remove
+	index := 0
+	for i := 0; i < len(consumerIds); i = i + 1 {
+		if consumerIds[i] == consumerId {
+			index = i
+			break
+		}
+	}
+	if consumerIds[index] != consumerId {
+		return fmt.Errorf("failed to find consumer id (%s) in the chains to be launched", consumerId)
+	}
+
+	if len(consumerIds) == 1 {
+		store.Delete(types.SpawnTimeToConsumerIdsKey(spawnTime))
+		return nil
+	}
+
+	updatedConsumerIds := append(consumerIds[:index], consumerIds[index+1:]...)
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	err = enc.Encode(updatedConsumerIds)
+	if err != nil {
+		return err
+	}
+
+	store.Set(types.SpawnTimeToConsumerIdsKey(spawnTime), buf.Bytes())
+	return nil
+}
+
+// TODO (PERMISSIONLESS) merge the functions, they practicall do the same
+
+// GetConsumersToBeStopped
+func (k Keeper) GetConsumersToBeStopped(ctx sdk.Context, stopTime time.Time) ([]string, error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.StopTimeToConsumerIdsKey(stopTime))
+	if bz == nil {
+		return []string{}, nil
+	}
+
+	var consumerIds []string
+	buf := bytes.NewBuffer(bz)
+	dec := gob.NewDecoder(buf)
+	err := dec.Decode(&consumerIds)
+	return consumerIds, err
+}
+
+// AppendSpawnTimeForConsumerToBeStopped
+func (k Keeper) AppendStopTimeForConsumerToBeStopped(ctx sdk.Context, consumerId string, stopTime time.Time) error {
+	store := ctx.KVStore(k.storeKey)
+
+	consumerIds, err := k.GetConsumersToBeStopped(ctx, stopTime)
+	if err != nil {
+		return err
+	}
+	consumerIds = append(consumerIds, consumerId)
+
+	// sort so that we avoid getting a consumer id in front of everyone else and delying the removal of a chain
+	sort.Slice(consumerIds, func(i, j int) bool {
+		return consumerIds[i] < consumerIds[j]
+	})
+
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	err = enc.Encode(consumerIds)
```

Comment:
Are you referring to the [slash acks](https://github.com/cosmos/interchain-security/blob/main/x/ccv/provider/keeper/keeper.go#L560) in the `keeper.go`? We can use proto un/marshalling there because all the `[]string`s is in the proto-generated `struct` `SlashAcks`.
I do not see a benefit in creating a proto-generated struct `ConsumerIds`. Am I missing something?

Reply:
@sainoe @MSalopek what do you guys think about this?

Reply:
My opinion is to use either `json` or `gob` formats. I have a slight preference for the former as it's widely used, including in the SDK.

Reply:
To remain in sync with the rest of the ecosystem please define a proto struct and use it.

Something like this should do the trick.

```proto
message ConsumerIds {
  repeated string ids;
}
```